cnxdb/migrations/20170911201035_transform_cnxml_to_html.py

```diff
@@ -0,0 +1,113 @@
+# -*- coding: utf-8 -*-
+import logging
+import time
+from datetime import timedelta
+
+import rhaptos.cnxmlutils
+from dbmigrator import deferred, logger
+
+
+def _batcher(seq, size):
+    for pos in range(0, len(seq), size):
+         yield seq[pos:pos + size]
+
+
+def should_run(cursor, limit='LIMIT 1'):
+    version = rhaptos.cnxmlutils.__version__
+    version_text = '%data-cnxml-to-html-ver="{}"%'.format(version)
+    logger.info('Looking for {}'.format(version_text))
+    cursor.execute("""\
+SELECT files.fileid, max(module_ident)
+    FROM module_files
+    NATURAL JOIN files
+    WHERE filename = 'index.cnxml.html'
+      AND convert_from(file, 'utf-8') NOT LIKE %s
+    GROUP BY files.fileid {}""".format(limit),
+                   (version_text,))
+    return cursor.fetchall()
+
+
+@deferred
+def up(cursor):
+    """Transform all index.cnxml to index.cnxml.html"""
+    # Get all the index.cnxml.html that does not have rhaptos.cnxmlutils
+    # current version
+    to_transform = should_run(cursor, limit='')
+    num_todo = len(to_transform)
+
+    batch_size = 10
+    logger.info('Batch size: {}'.format(batch_size))
+
+    start = time.time()
+    guesstimate = 1 * num_todo
+    guess_complete = guesstimate + start
+    logger.info('Completion guess: '
+                '"{}" ({})'.format(time.ctime(guess_complete),
+                                   timedelta(0, guesstimate)))
+
+    cursor.execute('DROP TABLE IF EXISTS datamigrations.index_cnxml_html')
```

Comment:
Hmm, was hoping we could do some sort of restart-able conversion, but leave rollback available. This drop table will stomp on any previous rollback

Reply:
It's a bit hard to decide when to clear the `datamigrations.index_cnxml_html` table...  What do you think?

Reply:
I think that's out-of-scope for migrations - operations clears it out later. That's what bit me with the simple automated script - I used  a temp table to 'automatically cleanup' then ya'll both needed that data for a rollback :-( (hmm, perhaps we need a new class of migrations - explicit only, that only run if you give the date-hash …)

Reply:
I suggest insert into the table if it's there, otherwise create it.